File: agent/integrations/opensandbox.py
"""OpenSandbox backend implementation."""

from __future__ import annotations

import asyncio
import logging
import os
import threading
from typing import Any

from deepagents.backends.protocol import (
    ExecuteResponse,
    FileDownloadResponse,
    FileUploadResponse,
    SandboxBackendProtocol,
    WriteResult,
)
from deepagents.backends.sandbox import BaseSandbox

logger = logging.getLogger(__name__)

DEFAULT_OPENSANDBOX_TEMPLATE = (
    "sandbox-registry.cn-zhangjiakou.cr.aliyuncs.com/opensandbox/code-interpreter:v1.0.2"
)
DEFAULT_OPENSANDBOX_TIMEOUT = 300
DEFAULT_OPENSANDBOX_URL = "http://localhost:9000"
ANTHROPIC_API_KEY_ENV = "ANTHROPIC_API_KEY"


def _combine_output(stdout: Any, stderr: Any) -> str:
    parts = [str(part) for part in (stdout, stderr) if part]
    return "\n".join(parts)


def _extract_execution_output(result: Any) -> tuple[str, str]:
    """Extract stdout and stderr text from an opensandbox Execution object."""
    stdout_msgs = getattr(getattr(result, "logs", None), "stdout", [])
    stderr_msgs = getattr(getattr(result, "logs", None), "stderr", [])
    stdout = "".join(getattr(m, "text", "") for m in stdout_msgs)
    stderr = "".join(getattr(m, "text", "") for m in stderr_msgs)
    return stdout, stderr


def _parse_server_url(server_url: str) -> str:
    """Extract domain (host:port) from a server URL for use in ConnectionConfig."""
    from urllib.parse import urlparse

    parsed = urlparse(server_url)
    return parsed.netloc or server_url.replace("http://", "").replace("https://", "")


def _check_server_reachable(server_url: str) -> None:
    """Verify the OpenSandbox server is reachable via a quick TCP probe.

    Raises:
        ConnectionError: If the server cannot be reached, with a clear message
            that includes the configured URL and instructions to fix it.
    """
    import socket
    from urllib.parse import urlparse

    parsed = urlparse(server_url)
    host = parsed.hostname or "localhost"
    port = parsed.port or 9000
    try:
        with socket.create_connection((host, port), timeout=3):
            pass
    except OSError as exc:
        raise ConnectionError(
            f"OpenSandbox server is not reachable at {server_url} "
            f"(host={host}, port={port}). "
            "Ensure the server is running and OPENSANDBOX_URL is set correctly. "
            f"Original error: {exc}"
        ) from exc


def _build_sandbox_envs() -> dict[str, str]:
    api_key = os.environ.get(ANTHROPIC_API_KEY_ENV)
    if not api_key:
        return {}
    return {ANTHROPIC_API_KEY_ENV: api_key}


def _resolve_timeout(timeout: int | None) -> int:
    if timeout is not None:
        return timeout
    env_timeout = os.environ.get("OPENSANDBOX_TIMEOUT") or os.environ.get("SANDBOX_TIMEOUT")
    if env_timeout:
        try:
            return int(env_timeout)
        except ValueError:
            logger.warning(
                "Invalid OPENSANDBOX_TIMEOUT=%r, falling back to %s",
                env_timeout,
                DEFAULT_OPENSANDBOX_TIMEOUT,
            )
    return DEFAULT_OPENSANDBOX_TIMEOUT


class OpenSandboxBackend(BaseSandbox):
    """OpenSandbox backend implementation conforming to SandboxBackendProtocol.

    All async operations (including sandbox creation/connection) run in a single
    dedicated background event loop so that the SDK's internal httpx transports
    are always used on the loop they were created on, preventing "Event loop is
    closed" errors from cross-loop transport cleanup.
    """

    def __init__(
        self,
        *,
        sandbox_id: str | None = None,
        connection_config: Any,
        create_kwargs: dict[str, Any] | None = None,
        default_timeout: int = DEFAULT_OPENSANDBOX_TIMEOUT,
    ) -> None:
        self._connection_config = connection_config
        self._default_timeout = default_timeout
        # Start a dedicated background event loop.  All SDK calls (including the
        # ones that create httpx transports) happen on this loop.
        self._loop = asyncio.new_event_loop()
        self._loop_thread = threading.Thread(target=self._loop.run_forever, daemon=True)
        self._loop_thread.start()
        # Create or reconnect to the sandbox inside the dedicated loop.
        self._sandbox = asyncio.run_coroutine_threadsafe(
            self._init_sandbox(sandbox_id, create_kwargs or {}),
            self._loop,
        ).result()

    async def _init_sandbox(self, sandbox_id: str | None, create_kwargs: dict[str, Any]) -> Any:
        from opensandbox import Sandbox

        if sandbox_id:
            return await Sandbox.connect(sandbox_id, connection_config=self._connection_config)
        return await Sandbox.create(connection_config=self._connection_config, **create_kwargs)

    def _run_async(self, coro: Any) -> Any:
        """Submit a coroutine to the dedicated background event loop and block."""
        future = asyncio.run_coroutine_threadsafe(coro, self._loop)
        return future.result()

    @property
    def id(self) -> str:
        sandbox_id = getattr(self._sandbox, "id", None)
        if not sandbox_id:
            msg = "OpenSandbox did not expose an id"
            raise AttributeError(msg)
        return sandbox_id

    def execute(self, command: str, *, timeout: int | None = None) -> ExecuteResponse:
        effective_timeout = timeout if timeout is not None else self._default_timeout

        async def _execute() -> ExecuteResponse:
            from datetime import timedelta

            from opensandbox.models.execd import RunCommandOpts

            opts = RunCommandOpts(timeout=timedelta(seconds=effective_timeout))
            try:
                result = await self._sandbox.commands.run(command, opts=opts)
            except Exception as exc:
                return ExecuteResponse(output=str(exc), exit_code=1, truncated=False)

            stdout, stderr = _extract_execution_output(result)
            output = _combine_output(stdout, stderr)
            exit_code = getattr(result, "exit_code", None)
            if exit_code is None:
                exit_code = 0
            return ExecuteResponse(
                output=output,
                exit_code=exit_code,
                truncated=False,
            )

        return self._run_async(_execute())

    def write(self, file_path: str, content: str) -> WriteResult:
        """Write content using the OpenSandbox SDK to avoid ARG_MAX."""
        try:

            async def _write() -> WriteResult:
                from opensandbox.models import WriteEntry

                entry = WriteEntry(path=file_path, data=content.encode("utf-8"), mode=644)
                await self._sandbox.files.write_files([entry])
                return WriteResult(path=file_path, files_update=None)

            return self._run_async(_write())
        except Exception as exc:
            return WriteResult(path=file_path, error=f"Failed to write file '{file_path}': {exc}")

    def download_files(self, paths: list[str]) -> list[FileDownloadResponse]:
        """Download multiple files from the OpenSandbox sandbox."""

        async def _download() -> list[FileDownloadResponse]:
            responses: list[FileDownloadResponse] = []
            for path in paths:
                try:
                    content = await self._sandbox.files.read_file(path)
                    # SDK may return str or bytes; protocol expects bytes.
                    if isinstance(content, str):
                        content = content.encode("utf-8")
                    responses.append(FileDownloadResponse(path=path, content=content, error=None))
                except Exception as exc:
                    responses.append(FileDownloadResponse(path=path, content=None, error=str(exc)))
            return responses

        return self._run_async(_download())

    def upload_files(self, files: list[tuple[str, bytes]]) -> list[FileUploadResponse]:
        """Upload multiple files to the OpenSandbox sandbox."""

        async def _upload() -> list[FileUploadResponse]:
            from opensandbox.models import WriteEntry

            responses: list[FileUploadResponse] = []
            entries = [WriteEntry(path=path, data=data, mode=644) for path, data in files]
            try:
                await self._sandbox.files.write_files(entries)
                for path, _ in files:
                    responses.append(FileUploadResponse(path=path, error=None))
            except Exception as exc:
                for path, _ in files:
                    responses.append(FileUploadResponse(path=path, error=str(exc)))
            return responses

        return self._run_async(_upload())


def _update_thread_sandbox_metadata(sandbox_id: str) -> None:
    """Update thread metadata with sandbox_id."""
    try:
        from langgraph.config import get_config
        from langgraph_sdk import get_client

        config = get_config()
        thread_id = config.get("configurable", {}).get("thread_id")
        if not thread_id:
            return
        client = get_client()

        async def _update() -> None:
            await client.threads.update(
                thread_id=thread_id,
                metadata={"sandbox_id": sandbox_id},
            )

        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            asyncio.run(_update())
        else:
            loop.create_task(_update())
    except Exception:
        pass


def create_opensandbox_sandbox(
    sandbox_id: str | None = None,
    timeout: int | None = None,
) -> SandboxBackendProtocol:
    """Create or connect to an OpenSandbox sandbox.

    Args:
        sandbox_id: Optional existing sandbox ID to reconnect to.
        timeout: Optional timeout for the sandbox lease in seconds.

    Returns:
        An OpenSandbox backend implementing SandboxBackendProtocol.

    Raises:
        ImportError: If opensandbox-code-interpreter is not installed.
        ValueError: If required environment variables are not set.
    """
    try:
        from datetime import timedelta

        from opensandbox.config import ConnectionConfig
    except ImportError as exc:
        raise ImportError(
            "OpenSandbox requires the 'opensandbox-code-interpreter' package. "
            "Install it with: uv sync --extra opensandbox "
            "or uv pip install 'opensandbox-code-interpreter'"
        ) from exc

    template = os.environ.get("OPENSANDBOX_TEMPLATE", DEFAULT_OPENSANDBOX_TEMPLATE)
    envs = _build_sandbox_envs()
    resolved_timeout = _resolve_timeout(timeout)
    server_url = os.environ.get("OPENSANDBOX_URL", DEFAULT_OPENSANDBOX_URL)

    # Fail fast with a clear message before spinning up the background loop.
    _check_server_reachable(server_url)

    # Build connection config: extract domain from server_url and enable server proxy
    # so execd requests are routed through the sandbox server (needed for Docker setups
    # where the client cannot reach the sandbox container directly).
    domain = _parse_server_url(server_url) if server_url else None
    connection_config = ConnectionConfig(domain=domain, use_server_proxy=True)

    create_kwargs: dict[str, Any] | None = None
    if not sandbox_id:
        from opensandbox.models.sandboxes import NetworkPolicy

        # By default, do NOT attach the egress sidecar (network_policy=None).
        # Without a sidecar the sandbox uses plain Docker bridge networking, which
        # provides full outbound internet access via Docker's built-in NAT — this
        # is required for the agent to browse the web, install packages, and use
        # Playwright.  The egress sidecar intercepts DNS via iptables and can break
        # browser/playwright connections even in allow-all mode.
        #
        # Set OPENSANDBOX_DENY_EGRESS=true to attach the sidecar with a deny-all
        # egress policy (requires [egress] image to be configured in ~/.sandbox.toml
        # and docker.network_mode = "bridge").
        deny_egress = os.environ.get("OPENSANDBOX_DENY_EGRESS", "").lower() in (
            "1",
            "true",
            "yes",
        )
        network_policy: NetworkPolicy | None = (
            NetworkPolicy(default_action="deny") if deny_egress else None
        )

        create_kwargs = {
            "image": template,
            "timeout": timedelta(seconds=resolved_timeout),
            "network_policy": network_policy,
        }
        if envs:
            create_kwargs["env"] = envs

    backend = OpenSandboxBackend(
        sandbox_id=sandbox_id,
        connection_config=connection_config,
        create_kwargs=create_kwargs,
        default_timeout=resolved_timeout,
    )
    _update_thread_sandbox_metadata(backend.id)
    return backend

File: src/integrations/opensandbox.ts
/**
 * OpenSandbox integration — self-hosted sandbox provider.
 * Mirrors agent/integrations/opensandbox.py
 *
 * OpenSandbox is an open-source sandbox runtime from Alibaba:
 * https://github.com/alibaba/OpenSandbox
 *
 * Env vars:
 *   OPENSANDBOX_URL       — server URL (default: http://localhost:9000)
 *   OPENSANDBOX_TEMPLATE  — container image (default: opensandbox/code-interpreter:v1.0.2)
 *   OPENSANDBOX_TIMEOUT   — default timeout in seconds (default: 300)
 *   OPENSANDBOX_DENY_EGRESS — set "true" to block all outbound traffic
 *   ANTHROPIC_API_KEY     — forwarded into sandbox as env var if set
 */

import type {
  SandboxBackendProtocol,
  ExecuteResponse,
  WriteResult,
  FileInfo,
  GrepMatch,
  FileDownloadResponse,
  FileUploadResponse,
  FileOperationError,
} from "deepagents";

const DEFAULT_OPENSANDBOX_URL = "http://localhost:9000";
const DEFAULT_OPENSANDBOX_TEMPLATE =
  "sandbox-registry.cn-zhangjiakou.cr.aliyuncs.com/opensandbox/code-interpreter:v1.0.2";
const DEFAULT_OPENSANDBOX_TIMEOUT = 300;

// ─── Helpers ──────────────────────────────────────────────────────────────────

function resolveTimeout(timeout?: number): number {
  if (timeout !== undefined) return timeout;
  const env = process.env.OPENSANDBOX_TIMEOUT ?? process.env.SANDBOX_TIMEOUT;
  if (env) {
    const n = parseInt(env, 10);
    if (!isNaN(n)) return n;
  }
  return DEFAULT_OPENSANDBOX_TIMEOUT;
}

function buildSandboxEnvs(): Record<string, string> {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  return apiKey ? { ANTHROPIC_API_KEY: apiKey } : {};
}

/** Extract host:port from a URL string for ConnectionConfig. */
function parseServerDomain(serverUrl: string): string {
  try {
    const u = new URL(serverUrl);
    return u.host; // "host:port" or "host"
  } catch {
    return serverUrl.replace(/^https?:\/\//, "");
  }
}

import net from "node:net";

/** Verify the OpenSandbox server is reachable (TCP connect). */
async function checkServerReachable(serverUrl: string): Promise<void> {
  let host = "localhost";
  let port = 9000;
  try {
    const u = new URL(serverUrl);
    host = u.hostname;
    port = parseInt(u.port || "9000", 10);
  } catch {
    // keep defaults
  }

  return new Promise((resolve, reject) => {
    const socket = net.createConnection({ host, port }, () => {
      socket.destroy();
      resolve();
    });
    socket.setTimeout(3000);
    socket.on("error", (err: Error) => {
      reject(
        new Error(
          `OpenSandbox server is not reachable at ${serverUrl} (host=${host}, port=${port}). ` +
            `Ensure the server is running and OPENSANDBOX_URL is set correctly. ` +
            `Original error: ${err.message}`,
        ),
      );
    });
    socket.on("timeout", () => {
      socket.destroy();
      reject(
        new Error(
          `OpenSandbox server timed out at ${serverUrl} (host=${host}, port=${port}).`,
        ),
      );
    });
  });
}

// ─── OpenSandboxBackend ───────────────────────────────────────────────────────

/**
 * Wraps the opensandbox SDK in the SandboxBackendProtocol interface.
 *
 * The opensandbox SDK is imported dynamically so this file compiles even if
 * the optional package is not installed — an ImportError is only surfaced at
 * runtime when SANDBOX_TYPE=opensandbox.
 */
export class OpenSandboxBackend implements SandboxBackendProtocol {
  // biome-ignore lint/suspicious/noExplicitAny: opensandbox SDK types not available at compile time
  private readonly _sandbox: any;
  private readonly _defaultTimeout: number;

  // biome-ignore lint/suspicious/noExplicitAny: opensandbox SDK types not available at compile time
  constructor(sandbox: any, defaultTimeout: number) {
    this._sandbox = sandbox;
    this._defaultTimeout = defaultTimeout;
  }

  get id(): string {
    const sid = this._sandbox?.id;
    if (!sid) throw new Error("OpenSandbox did not expose an id");
    return String(sid);
  }

  // ─── Execute ───────────────────────────────────────────────────────────────

  async execute(command: string): Promise<ExecuteResponse> {
    try {
      // The opensandbox SDK exposes commands.run(cmd, opts)
      const result = await this._sandbox.commands.run(command, {
        timeout: this._defaultTimeout * 1000, // milliseconds
      });

      // Extract stdout/stderr from logs
      const stdoutMsgs: unknown[] = result?.logs?.stdout ?? [];
      const stderrMsgs: unknown[] = result?.logs?.stderr ?? [];
      const stdout = stdoutMsgs
        // biome-ignore lint/suspicious/noExplicitAny: dynamic SDK response
        .map((m: any) => m?.text ?? "")
        .join("");
      const stderr = stderrMsgs
        // biome-ignore lint/suspicious/noExplicitAny: dynamic SDK response
        .map((m: any) => m?.text ?? "")
        .join("");

      const output = [stdout, stderr].filter(Boolean).join("\n");
      const exitCode: number = result?.exit_code ?? 0;

      return { output, exitCode, truncated: false };
    } catch (err) {
      return {
        output: err instanceof Error ? err.message : String(err),
        exitCode: 1,
        truncated: false,
      };
    }
  }

  // ─── File operations ───────────────────────────────────────────────────────

  async read(filePath: string, offset = 0, limit = 500): Promise<string> {
    try {
      const result = await this.execute(
        `sed -n '${offset + 1},${offset + limit}p' ${shellQuote(filePath)}`,
      );
      return result.output;
    } catch (err) {
      return `Error reading file: ${err instanceof Error ? err.message : String(err)}`;
    }
  }

  // biome-ignore lint/suspicious/noExplicitAny: generic FileData
  async readRaw(filePath: string): Promise<any> {
    const text = await this.read(filePath);
    return { content: new TextEncoder().encode(text), mime_type: "text/plain" };
  }

  async write(filePath: string, content: string): Promise<WriteResult> {
    try {
      // Use the SDK's file write if available
      if (this._sandbox?.files?.write_files) {
        await this._sandbox.files.write_files([
          { path: filePath, data: new TextEncoder().encode(content), mode: 0o644 },
        ]);
        return { path: filePath, error: undefined };
      }
      // Fallback: use tee via execute
      const encoded = Buffer.from(content).toString("base64");
      const result = await this.execute(
        `echo ${shellQuote(encoded)} | base64 -d > ${shellQuote(filePath)}`,
      );
      if (result.exitCode !== 0) {
        return { path: filePath, error: result.output };
      }
      return { path: filePath, error: undefined };
    } catch (err) {
      return {
        path: filePath,
        error: `Failed to write file '${filePath}': ${err instanceof Error ? err.message : String(err)}`,
      };
    }
  }

  async edit(
    filePath: string,
    oldString: string,
    newString: string,
    replaceAll = false,
  // biome-ignore lint/suspicious/noExplicitAny: EditResult from deepagents
  ): Promise<any> {
    const content = await this.read(filePath);
    const occurrences = content.split(oldString).length - 1;
    if (occurrences === 0) {
      return { error: `String not found in ${filePath}`, path: filePath, occurrences: 0 };
    }
    const updated = replaceAll
      ? content.split(oldString).join(newString)
      : content.replace(oldString, newString);
    const writeResult = await this.write(filePath, updated);
    return {
      error: writeResult.error,
      path: filePath,
      occurrences: replaceAll ? occurrences : 1,
    };
  }

  async lsInfo(path: string): Promise<FileInfo[]> {
    const result = await this.execute(
      `ls -la ${shellQuote(path)} 2>&1 | tail -n +2`,
    );
    if (result.exitCode !== 0) return [];

    return result.output
      .split("\n")
      .filter(Boolean)
      .map((line) => {
        const parts = line.split(/\s+/);
        const name = parts[parts.length - 1] ?? "";
        const isDir = line.startsWith("d");
        return {
          path: `${path.replace(/\/$/, "")}/${name}${isDir ? "/" : ""}`,
          size: parseInt(parts[4] ?? "0", 10) || 0,
          is_dir: isDir,
        } satisfies FileInfo;
      });
  }

  async grepRaw(
    pattern: string,
    path?: string | null,
    glob?: string | null,
  ): Promise<GrepMatch[] | string> {
    const pathArg = path ? shellQuote(path) : ".";
    const globArg = glob ? `--include=${shellQuote(glob)}` : "";
    const result = await this.execute(
      `grep -rn ${globArg} ${shellQuote(pattern)} ${pathArg} 2>/dev/null`,
    );
    if (result.exitCode !== 0) return [];

    return result.output
      .split("\n")
      .filter(Boolean)
      .map((line) => {
        const [file, lineNum, ...rest] = line.split(":");
        return {
          path: file ?? "",
          line: parseInt(lineNum ?? "0", 10),
          text: rest.join(":"),
        } satisfies GrepMatch;
      });
  }

  async globInfo(pattern: string, path = "/"): Promise<FileInfo[]> {
    const result = await this.execute(
      `find ${shellQuote(path)} -path ${shellQuote(pattern)} 2>/dev/null`,
    );
    if (result.exitCode !== 0) return [];

    return result.output
      .split("\n")
      .filter(Boolean)
      .map((p) => ({
        path: p,
        size: 0,
        is_dir: false,
        last_modified: "",
        mime_type: "application/octet-stream",
      }));
  }

  // Optional bulk file operations
  async downloadFiles(paths: string[]): Promise<FileDownloadResponse[]> {
    return Promise.all(
      paths.map(async (p) => {
        try {
          if (this._sandbox?.files?.read_file) {
            const content = await this._sandbox.files.read_file(p);
            return {
              path: p,
              content: typeof content === "string" ? new TextEncoder().encode(content) : content,
              error: null,
            } satisfies FileDownloadResponse;
          }
          const result = await this.read(p);
          return { path: p, content: new TextEncoder().encode(result), error: null } satisfies FileDownloadResponse;
        } catch {
          return { path: p, content: null, error: "file_not_found" as FileOperationError } satisfies FileDownloadResponse;
        }
      }),
    );
  }

  async uploadFiles(files: Array<[string, Uint8Array]>): Promise<FileUploadResponse[]> {
    return Promise.all(
      files.map(async ([path, data]) => {
        const content = new TextDecoder().decode(data);
        const result = await this.write(path, content);
        return {
          path,
          error: result.error ? ("permission_denied" as FileOperationError) : null,
        } satisfies FileUploadResponse;
      }),
    );
  }
}

// ─── Factory ──────────────────────────────────────────────────────────────────

/**
 * Create or reconnect to an OpenSandbox sandbox.
 */
export async function createOpenSandbox(
  sandboxId?: string,
  timeout?: number,
): Promise<SandboxBackendProtocol> {
  let Sandbox: { create: Function; connect: Function };
  // biome-ignore lint/suspicious/noExplicitAny: dynamic import
  let ConnectionConfig: any;

  try {
    // biome-ignore lint/suspicious/noExplicitAny: optional peer dependency
    const sdk = await import("opensandbox" as any);
    Sandbox = sdk.Sandbox;
    ConnectionConfig = sdk.ConnectionConfig ?? sdk.config?.ConnectionConfig;
  } catch {
    throw new Error(
      "OpenSandbox requires the 'opensandbox' npm package. " +
        "Install it with: pnpm add opensandbox",
    );
  }

  const serverUrl = process.env.OPENSANDBOX_URL ?? DEFAULT_OPENSANDBOX_URL;
  const template = process.env.OPENSANDBOX_TEMPLATE ?? DEFAULT_OPENSANDBOX_TEMPLATE;
  const resolvedTimeout = resolveTimeout(timeout);
  const envs = buildSandboxEnvs();

  // Fail fast if server is not reachable
  await checkServerReachable(serverUrl);

  const domain = parseServerDomain(serverUrl);
  const connectionConfig = new ConnectionConfig({ domain, useServerProxy: true });

  // biome-ignore lint/suspicious/noExplicitAny: dynamic SDK
  let rawSandbox: any;

  if (sandboxId) {
    rawSandbox = await Sandbox.connect(sandboxId, { connection_config: connectionConfig });
  } else {
    const denyEgress = ["1", "true", "yes"].includes(
      (process.env.OPENSANDBOX_DENY_EGRESS ?? "").toLowerCase(),
    );

    const createOpts: Record<string, unknown> = {
      image: template,
      timeout: resolvedTimeout * 1000, // ms
      connection_config: connectionConfig,
    };

    if (denyEgress) {
      createOpts.network_policy = { default_action: "deny" };
    }

    if (Object.keys(envs).length) {
      createOpts.env = envs;
    }

    rawSandbox = await Sandbox.create(createOpts);
  }

  return new OpenSandboxBackend(rawSandbox, resolvedTimeout) as unknown as SandboxBackendProtocol;
}

// ─── Internal ─────────────────────────────────────────────────────────────────

function shellQuote(s: string): string {
  return `'${s.replace(/'/g, "'\\''")}'`;
}
